### ISSUE TITLE
consider an ubuntu or k3os base OS for running k3s

### ISSUE THREAD

**Issue by billimek** [billimek/homelab-infrastructure#19]
Previously I was running k3s installed manually on top of ubuntu-based VMs, nodes, and pi4s.

Recently I experimented and eventually migrated fully to k3os for a few reasons:

* It is extremely simple and minimal designed for only running k3s
* It is much easier to make the nodes 'cattle' instead of 'pets' due to the declarative nature of the configuration and installation
* Upgrades are completely built-in via the system upgrade controller

However, there were some few downsides as well:

* **VLAN Configuration**: It wasn't straight-forward or possible to declare VLAN membership within the node configuration itself because of the use of connman. It _may_ be possible to hack something to make connman work as a declarative vlan configuration but it looks ugly
* **Odroid H2**: The k3os ISO installer is based on arch and requires EFI.  There are apparently some issues installing a EFI based linux system on the odroid.  I wrestled with this for about half a day before giving up on installing k3os via ISO.  It was ultimately possible to install something like ubuntu first and then install k3os vis the overlay method, but that has problems too (see below)
* **rbd kernel module on arm64**:  In order to consume ceph storage by kubernetes workloads running on arm64-based nodes, the host OS needs to provide the `rbd` kernel module.  The existing excellent [picl-k3os-image-generator](https://github.com/sgielen/picl-k3os-image-generator) is based only on rasbian which does not ship with the rbd module.  I spent a while trying to find one to inject or add, but could only find it in the existing ubuntu arm64 OSs and the kernel versions were wildly different.  It was possible to solve the problem by installing ubuntu arm64 on the pi4 and then run the k3os overlay installation, but that has problems too (see below)
* **k3os overlay installation open file descriptor issue**: This is ultimately what led me to reconsider running a k3os-based cluster.  There is a lot of detail in a related [issue](https://github.com/rancher/k3s/issues/1706) and a resolution is yet to be discovered.  Basically,  for some reason, running k3os via overlay on either arm64 or amd64 ubuntu base OSs presents a problem with a liner growth in open file descriptors that can't be overcome unless there is a reboot.  The rate of growth results in a locked node after about a week.  At present it is not known what the root cause to this is.

It is possible to address the last issue by simply running the odroid and pi4s as regular ubuntu OSs with k3s installed directly on them.  But this results in a cluster with nodes created via different approaches.  This also leads to issues with the system upgrade controller based on my experience.

Using the cloud init and runcmd approach with ubuntu server, it's doable to create a unique ISO for each node that will auto install k3s and automatically join the cluster.  I would also like to explore installing ubuntu server via netboot that there is a lot of [drama](https://discourse.ubuntu.com/t/netbooting-the-live-server-installer/14510) in that [space](https://discourse.ubuntu.com/t/server-installer-plans-for-20-04-lts/13631) right now, so it's unclear what the proper path forward is for netbooting modern versions of ubuntu.

Ultimately, I believe that the nodes in the cluster should adopt a consistent OS installation.  This issue is to collect information on the two possible approaches (or maybe another approach all-together) instead of a bunch of open tabs.

**Comment by billimek** [comment]
Regarding bootstrapping a new ubuntu-based k3s node, the following cloud-init yaml [example](https://github.com/billimek/homelab-infrastructure/blob/master/k3os/arm64/nodes/k3s-pi4-a) seems to do the trick, the pertinent part is in the `runcmd` section:

```yaml
runcmd:
- ln -sfn /run/systemd/resolve/resolv.conf /etc/resolv.conf
- curl -sfL https://get.k3s.io | K3S_URL=https://k3s-0:6443 K3S_TOKEN=$K3OS_TOKEN sh -s - --node-taint arm=true:NoExecute 
- sync
- reboot -f
```

... this ultimately results in a single 'ISO' that will update properly, set the proper password and access controls, install k3s, and automatically join the node.  This sort of simulates the initial experience of running k3os but requires a massive 900MB ISO (ubuntu server images are now ridiculously bloated IMO) to get there.

**Comment by billimek** [comment]
Also consider [alpine linux](https://alpinelinux.org/downloads/) as a minimal OS for running k3s.  [This](https://teada.net/k3s-on-alpine-linux/) is a basic rundown of manually installing k3s.

Brief searching suggests that rbd is an option even for arm64 (aarch64)

**Comment by billimek** [comment]
Going to experiment more with all nodes being based on ubuntu.  This is the 'plan':

## Installation

* Attempt netboot
* If that isn't an option, attempt to use a generic ISO/image with cloud-init referenced somewhere on the network

## Configuration

* ALL configuration will be handled via cloud-init settings
  * ssh keys
  * default password
  * install baseline packages
  * set ntp to router
  * manually set DNS to router (or maybe just use dhcp)
  * configure vlan client-side
  * fix `/etc/resolv.conf` file symlink
* Install or join k3s cluster using cloud-init runcmd stanza

Basically, it should never be necessary to ssh to the node if I do this correctly

## "Day 2" operations

* Install system-upgrade-controller `Plan` to handle k3s upgrades
* Configure ubuntu to handle ubuntu security updates (I think it comes this way out of the box) - and cordon/reboot via `kured`
* Explore automated system & kernel updates with a `Plan` (e.g. `sudo apt dist-upgrade`) - not sure if this is a good idea to automate - need to think about this more

**Comment by billimek** [comment]
This looks like a really good cookbook for what I want to try to do: https://askubuntu.com/questions/1238070/deploy-ubuntu-20-04-on-bare-metal-or-virtualbox-vm-by-pxelinux-cloud-init-doesn

(also [this](https://askubuntu.com/questions/1235723/automated-20-04-server-installation-using-pxe-and-live-server-image) for more background & details)

**Comment by billimek** [comment]
Don't think I'm going to mess with netboot or pxe stuff.  Just create unique ubuntu ISOs for each node with a script.  Less fuss that way and it's a pattern I know well.

It should still handle the **Configuration** stuff referenced above.

**Comment by billimek** [comment]
Consider leveraging [matchbox](https://matchbox.psdn.io/) to handle iPXE/netboot provisioning.  Also see [this](https://github.com/poseidon/typhoon/blob/master/docs/topics/hardware.md#isc-dhcp) for hints about configuring edgerouter to use matchbox.

**Comment by billimek** [comment]
https://github.com/beldzhang/netboot/releases/tag/0.96.4702 looks interesting as well.

**Comment by billimek** [comment]
Decided against netboot/pxe/matchbox style booting and will instead use a heterogeneous approach, for now, to install ubuntu server on the nodes:

* VMs will leverage [ubuntu cloudimg OS images](https://cloud-images.ubuntu.com/focal/current/) in conjunction with a curated cloud-init mounted as a cdrom device
* Raspberry Pi4's will install like before with the [preinstalled server image](http://cdimage.ubuntu.com/releases/focal/release/ubuntu-20.04-preinstalled-server-arm64+raspi.img.xz) and required cloud-init copied to the SD card prior to first boot.
* Physical devices (odroid-h2) will be totally manual installation

More details about all of this in a pending PR.
